ibm_cloud_sdk_core/token_manager.py

```diff
@@ -0,0 +1,224 @@
+
+            with self.lock:
+                request_active = self.request_time > (current_time - 60)
+                if not request_active:
+                    self.request_time = current_time
+
+            if not request_active:
+                token_response = self.request_token()
+                self._save_token_info(token_response)
+                self.request_time = 0
+                return
+
+            # Sleep for 0.5 seconds before checking token again
+            time.sleep(0.5)
+
+    @abstractmethod  # pragma: no cover
+    def request_token(self) -> None:
+        """Should be overridden by child classes."""
+        pass
+
+    @staticmethod
+    def _get_current_time() -> int:
+        return int(time.time())
+
+    def _is_token_expired(self) -> bool:
+        """
+        Check if currently stored token is expired.
+
+        Returns
+        -------
+        bool
+            True if token is expired; False otherwise
+        """
+        current_time = self._get_current_time()
+        return self.expire_time < current_time
+
+    def _token_needs_refresh(self) -> bool:
+        """
+        Check if currently stored token needs refresh.
+
+        Returns
+        -------
+        bool
+            True if token needs refresh; False otherwise
+        """
+        current_time = self._get_current_time()
+
+        with self.lock:
+            needs_refresh = self.refresh_time < current_time
+            if needs_refresh:
+                self.refresh_time = current_time + 60
+
+        return needs_refresh
+
+    @abstractmethod
+    def _save_token_info(self, token_response: dict) -> None:  # pragma: no cover
+        """
+        Decode the access token and save the response from the service to the object's state
+
+        Refresh time is set to approximately 80% of the token's TTL to ensure that
+        the token refresh completes before the current token expires.
+
+        Parameters
+        ----------
+        token_response : dict
+            Response from token service
+        """
+        pass
+
+    def _request_raw(self,
+                     method,
+                     url,
+                     *,
+                     headers=None,
+                     params=None,
+                     data=None,
+                     auth_tuple=None,
+                     **kwargs):
+        kwargs = dict({"timeout": 60}, **kwargs)
+        kwargs = dict(kwargs, **self.http_config)
+
+        if self.disable_ssl_verification:
+            kwargs['verify'] = False
+
+        response = requests.request(
+            method=method,
+            url=url,
+            headers=headers,
+            params=params,
+            data=data,
+            auth=auth_tuple,
+            **kwargs)
+        if 200 <= response.status_code <= 299:
+            return response
+
+        raise ApiException(response.status_code, http_response=response)
+
+    def _set_expire_and_refresh_time(self, exp, ttl):
```

Comment:
I'm not sure we need this method.  Subclasses could simply implement this directly in their `_save_token_info` methods.  But if we leave it, it should have a doc string that describes its intended behavior.

Reply:
Inlined in 949f389